### PR TITLE
Bind image to a widget

### PR DESCRIPTION
**Pull request by MatthiasWM**
So the widget will release/delete the images when the widget itself is deleted, see #3 .

### CODE REVIEW

**Comment by MatthiasWM** [comment]
Calling bind_image() with a new image will delete the old image if it was bound, and then set the new image, and bind that. If old and new image are the same, nothing happens.

Calling image() with a new image will delete the old image if it was bound, and set the new image without binding it. If old and new are the same, the image will not be deleted, but it will be unbound.

Calling image() or bind_image() with NULL will delete the old image if it was bound and not set a new image.

Shared images are handled.

File: src/Fl_Widget.cxx [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include <FL/Fl_Widget.H>
 #include <FL/Fl_Group.H>
 #include <FL/Fl_Tooltip.H>
+#include <FL/Fl_Shared_Image.H>
```
Comment (Albrecht-S) [inline src/Fl_Widget.cxx:21]: This should no longer be needed since the second commit doesn't use it.

Should be removed.

**Comment by Albrecht-S** [comment]
Code looks good (untested). I'm not sure (and don't have time to check) that all "special cases" mentioned in [this comment](https://github.com/fltk/fltk/pull/589#issuecomment-1345324777) are also documented. It's important for the user docs. Particularly "If old and new are the same, the image will not be deleted, but it will be unbound" might be unexpected (but it's OK).

Thank you for taking the original PR and making such a good solution (IMHO).

**Comment by wcout** [comment]
My thoughts, at  a quick glance:

- Isn't there missing an important part, namely the `release()` of the images in the widget destructor `~Fl_Widget()` , or does this happen somewhere else?

- And should
`````cxx
void bind_image(int f) { set_flag(IMAGE_BOUND); }
`````
not be:
`````cxx
void bind_image(int f) { f ? set_flag(IMAGE_BOUND) : clear_flag(IMAGE_BOUND); }
`````
or something like that?

Same with `bind_deimage()`.

**Comment by Albrecht-S** [comment]
@wcout Good points, thank you for reviewing the code!

**Comment by MatthiasWM** [comment]
> * Isn't there missing an important part, namely the `release()` of the images in the widget destructor `~Fl_Widget()` , or does this happen somewhere else?

`~Fl_Widget` is calling `image(NULL);` and `damage(NULL);` which will delete the images if bound to the widget.

> * And should
> 
> ```c++
> void bind_image(int f) { set_flag(IMAGE_BOUND); }
> ```
> 
> not be:
> 
> ```c++
> void bind_image(int f) { f ? set_flag(IMAGE_BOUND) : clear_flag(IMAGE_BOUND); }
> ```

Oops ;-) . it's fixed in the recent commit.

**Comment by MatthiasWM** [comment]
> Thank you for taking the original PR and making such a good solution (IMHO).

Ah, sure. It's fun.

**Pull request merged**